src/hotspot/share/opto/rangeinference.cpp

```diff
@@ -0,0 +1,769 @@
+  if (i1 == t2 || t2 == Type::TOP) {
+    return i1;
+  }
+  const CT* i2 = CT::try_cast(t2);
+  if (i2 != nullptr) {
+    if (!dual) {
+    // meet
+      return make(TypeIntPrototype<S, U>{{MIN2(i1->_lo, i2->_lo), MAX2(i1->_hi, i2->_hi)},
+                                         {MIN2(i1->_ulo, i2->_ulo), MAX2(i1->_uhi, i2->_uhi)},
+                                         {i1->_bits._zeros & i2->_bits._zeros, i1->_bits._ones & i2->_bits._ones}},
+                  MAX2(i1->_widen, i2->_widen), false);
+    }
+    // join
+    return make(TypeIntPrototype<S, U>{{MAX2(i1->_lo, i2->_lo), MIN2(i1->_hi, i2->_hi)},
+                                       {MAX2(i1->_ulo, i2->_ulo), MIN2(i1->_uhi, i2->_uhi)},
+                                       {i1->_bits._zeros | i2->_bits._zeros, i1->_bits._ones | i2->_bits._ones}},
+                MIN2(i1->_widen, i2->_widen), true);
+  }
+
+  assert(t2->base() != i1->base(), "");
+  switch (t2->base()) {          // Switch on original type
+  case Type::AnyPtr:                  // Mixing with oops happens when javac
+  case Type::RawPtr:                  // reuses local variables
+  case Type::OopPtr:
+  case Type::InstPtr:
+  case Type::AryPtr:
+  case Type::MetadataPtr:
+  case Type::KlassPtr:
+  case Type::InstKlassPtr:
+  case Type::AryKlassPtr:
+  case Type::NarrowOop:
+  case Type::NarrowKlass:
+  case Type::Int:
+  case Type::Long:
+  case Type::FloatTop:
+  case Type::FloatCon:
+  case Type::FloatBot:
+  case Type::DoubleTop:
+  case Type::DoubleCon:
+  case Type::DoubleBot:
+  case Type::Bottom:                  // Ye Olde Default
+    return Type::BOTTOM;
+  default:                      // All else is a mistake
+    i1->typerr(t2);
+    return nullptr;
+  }
+}
+template const Type* int_type_xmeet(const TypeInt* i1, const Type* t2,
+                                    const Type* (*make)(const TypeIntPrototype<jint, juint>&, int, bool), bool dual);
+template const Type* int_type_xmeet(const TypeLong* i1, const Type* t2,
+                                    const Type* (*make)(const TypeIntPrototype<jlong, julong>&, int, bool), bool dual);
+
+// Called in PhiNode::Value during CCP, monotically widen the value set, do so rigorously
+// first, after WidenMax attempts, if the type has still not converged we speed up the
+// convergence by abandoning the bounds
+template <class CT>
+const Type* int_type_widen(const CT* nt, const CT* ot, const CT* lt) {
+  using S = std::remove_const_t<decltype(CT::_lo)>;
+  using U = std::remove_const_t<decltype(CT::_ulo)>;
+
+  if (ot == nullptr) {
+    return nt;
+  }
+
+  // If new guy is equal to old guy, no widening
+  if (int_type_equal(nt, ot)) {
+    return ot;
+  }
+
+  // If old guy contains new, then we probably widened too far & dropped to
+  // bottom. Return the wider fellow.
+  if (int_type_subset(ot, nt)) {
+    return ot;
+  }
+
+  // Neither contains each other, weird?
+  // fatal("Integer value range is not subset");
+  // return this;
```

Comment:
Yes it is, but I don't want to remove it. I think it should be the case the `new_type` contains `old_type`, but apparently sometimes it is not the case and I don't understand it. Returning the bottom type instead of `limit_type` is also weird, too.

Reply:
Unless you want to investigate it this should be removed. File RFE to look on it later.
Also you can execute it only in debug VM by using #ifdef ASSERT and `assert()` instead of `fatal()`.

Reply:
Done! I don't want to risk breaking anything so I will take a look later.